CesiumGS/cesium
# Add option to select which geocoder is used

# Description

Adds the ability to change the geocoder used.



Both Bing and Google have legal requirements that their geocoders should not be used with other data.  Eg. Using  Bing geocode results while showing Google 3D Tiles is not allowed.



- Adds `GoogleGeocoderService` for standalone usage of Google geocoder
- Adds `IonGeocodeProvider` enum to list the possible geocode providers when using Cesium ion.
- Updates `Viewer` constructor to also take an `IonGeocoderProvider` value
- Updates Sandcastles to specify Google or Bing geocoders where appropriate,  or remove the geocoder if other map data is used.

## Testing plan

1. Run `npm run start`
2. Open http://localhost:8080/Apps/Sandcastle/index.html
3. Type in _Philadelphia_ in the geocoder
    - Verify that multiple results are returned for Philadelphia
4. Open http://localhost:8080/Apps/Sandcastle/index.html?src=Google%20Photorealistic%203D%20Tiles%20with%20Building%20Insert.html
5. Type in _Philadelphia_ in the geocoder
    - Verify that only a single result is returned
    - Verify that there is only a single google attribution at the bottom of the screen
6. Open http://localhost:8080/Apps/Sandcastle/index.html?src=Bing%20Maps%20Labels%20Only.html
7. Type in _Philadelphia_ in the geocoder
    - Verify that multiple results are returned for Philadelphia
8. Run the following code snippets in the local sandcastle and verify the results.  Replace 'your-key-here` with a valid Google API key with Geocoding enabled

```js
const viewer = new Cesium.Viewer("cesiumContainer", {
  geocoder: new Cesium.GoogleGeocoderService({
    key: 'your-key-here'
  })
});
const googleTileset = await Cesium.createGooglePhotorealistic3DTileset({
  onlyUsingWithGoogleGeocoder: true
});
viewer.scene.primitives.add(googleTileset);
```
- Type _Philadelphia_ in the geocoder
- Verify that a single result appears and takes you to the correct location
- Verify that a single Google logo is shown on screen in the data attributions


```js
const viewer = new Cesium.Viewer("cesiumContainer");
const googleTileset = await Cesium.createGooglePhotorealistic3DTileset();
viewer.scene.primitives.add(googleTileset);
```
-  Verify there is a warning in the console about using the Google geocoder

```js
const viewer = new Cesium.Viewer("cesiumContainer");
const googleTileset = await Cesium.createGooglePhotorealistic3DTileset('your-key-here');
viewer.scene.primitives.add(googleTileset);
```

-  Verify that the tileset loads
- Verify there is a deprecation warning
```js
const viewer = new Cesium.Viewer("cesiumContainer");
const googleTileset = await Cesium.createGooglePhotorealistic3DTileset({
  key:'your-api-key-here', 
  onlyUsingWithGoogleGeocoder: true
});
viewer.scene.primitives.add(googleTileset);
```
- Verify that the tileset loads
- Verify that there is no deprecation warning.

# Author checklist

- [x] Update ion to use consistent Google attribution
- [x] Figure out test failure - https://github.com/CesiumGS/cesium/issues/12306 https://github.com/CesiumGS/cesium/pull/12307
- [x] Figure out one time warning
